### PR TITLE
Add percentage of ownership to eth2 validators

### PR DESCRIPTION
**Pull request by yabirgb**
Closes #(issue_number)

## Checklist

- [ ] The PR modified the frontend, and updated the [user guide](https://github.com/rotki/rotki/blob/develop/docs/usage_guide.rst) to reflect the changes.

### CODE REVIEW

**Comment by yabirgb** [comment]
From the chat yesterday in discord I assumed that the performance calculated in https://github.com/rotki/rotki/blob/e1eba5f80abda901160c8ffe9a38c336dcf9f32a/rotkehlchen/externalapis/beaconchain.py#L225 shouldn't be modified to take into account the % of ownership. Correct me if wrong

File: rotkehlchen/api/v1/resources.py [hunk 1]
```diff
@@ -1271,11 +1272,13 @@ def put(
             self,
             validator_index: Optional[int],
             public_key: Optional[Eth2PubKey],
+            ownership_percentage: FVal,
             async_query: bool,
     ) -> Response:
         return self.rest_api.add_eth2_validator(
             validator_index=validator_index,
             public_key=public_key,
+            ownership_proportion=ownership_percentage,
```
Comment (yabirgb) [inline rotkehlchen/api/v1/resources.py:1281]: This naming change from ownership_percentage to ownership_proportion is intended. I reused the filter we have for underlying assets where we pass the proportion using %. So for the reader to know what is being used in each place I named the variable from/to the frontend `ownership_percentage` and the one used in the backend (in the range 0-1) as `ownership_proportion`

File: rotkehlchen/api/v1/encoding.py [hunk 2]
```diff
@@ -2468,6 +2468,7 @@ class Eth2ValidatorSchema(Schema):
         ),
     )
     public_key = fields.String(load_default=None)
+    ownership_percentage = FloatingPercentageField(load_default=FVal(1))
```
Comment (LefterisJP) [inline rotkehlchen/api/v1/encoding.py:2471]: I  guess you tested this, but load_default skips the deserialization?

File: rotkehlchen/chain/ethereum/modules/eth2.py [hunk 3]
```diff
@@ -319,7 +321,12 @@ def get_validator_daily_stats(
         dbeth2 = DBEth2(self.database)
         return dbeth2.get_validator_daily_stats_and_limit_info(filter_query=filter_query)
 
-    def add_validator(self, validator_index: Optional[int], public_key: Optional[Eth2PubKey]) -> None:  # noqa: E501
+    def add_validator(
+        self,
+        validator_index: Optional[int],
+        public_key: Optional[Eth2PubKey],
+        ownership_proportion: FVal,
+    ) -> None:  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/eth2.py:329]: This can go away

```suggestion
    ) -> None:
```

File: rotkehlchen/chain/ethereum/typing.py [hunk 4]
```diff
@@ -108,9 +108,31 @@ def __hash__(self) -> int:
...
+    int,  # proposer_attester_slashings
+    int,  # deposits_number
+    str,  # amount_deposited
+    int,  # ownership_proportion
+]
+
 
 class ValidatorDailyStats(NamedTuple):
     validator_index: int  # keeping the index here so it can be shown in accounting
+    ownership_proportion: FVal  # Keeping this so it can be used during accounting
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/typing.py:135]: This is not needed. We already know the ownership proportion since we have it in the validator. We are not able to get, or interested to even try guessing historic ownership of validators
```suggestion
```

But that also means my query of summing PnL for the frontend that I just added in the other PR will need to become more complicated :sweat:

File: rotkehlchen/chain/ethereum/typing.py [hunk 5]
```diff
@@ -180,9 +203,10 @@ def to_db_tuple(self) -> ValidatorDailyStatsDBTuple:
         )
 
     @classmethod
-    def deserialize_from_db(cls, entry: ValidatorDailyStatsDBTuple) -> 'ValidatorDailyStats':
+    def deserialize_from_db(cls, entry: ValidatorDailyStatsDBTupleRead) -> 'ValidatorDailyStats':
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/typing.py:206]: not needed as I explained above. Also why did you add another completely new Tuple?

File: rotkehlchen/chain/ethereum/typing.py [hunk 6]
```diff
@@ -203,6 +227,7 @@ def deserialize_from_db(cls, entry: ValidatorDailyStatsDBTuple) -> 'ValidatorDai
     def serialize(self) -> Dict[str, Any]:
         return {
             'validator_index': self.validator_index,
+            'ownership_percentage': str(self.ownership_proportion * 100),
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/typing.py:230]: not needed

File: rotkehlchen/chain/ethereum/typing.py [hunk 7]
```diff
@@ -180,9 +203,10 @@ def to_db_tuple(self) -> ValidatorDailyStatsDBTuple:
         )
 
     @classmethod
-    def deserialize_from_db(cls, entry: ValidatorDailyStatsDBTuple) -> 'ValidatorDailyStats':
+    def deserialize_from_db(cls, entry: ValidatorDailyStatsDBTupleRead) -> 'ValidatorDailyStats':
         return cls(
             validator_index=entry[0],
+            ownership_proportion=FVal(entry[16]),
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/typing.py:209]: not needed

File: rotkehlchen/db/eth2.py [hunk 8]
```diff
@@ -167,7 +167,11 @@ def get_validator_daily_stats(
         """Gets all DB entries for validator daily stats according to the given filter"""
         cursor = self.db.conn.cursor()
         query, bindings = filter_query.prepare()
-        query = 'SELECT * from eth2_daily_staking_details ' + query
+        query = """
```
Comment (LefterisJP) [inline rotkehlchen/db/eth2.py:170]: What are you doing here? You do not need to touch this query.

File: rotkehlchen/db/filtering.py [hunk 9]
```diff
@@ -351,7 +351,7 @@ def prepare(self) -> Tuple[List[str], List[Any]]:
         if self.validators is None:
             return [], []
         questionmarks = '?' * len(self.validators)
-        return [f'validator_index IN ({",".join(questionmarks)})'], self.validators
+        return [f'eth2_daily_staking_details.validator_index IN ({",".join(questionmarks)})'], self.validators  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/db/filtering.py:354]: You do not need to touch this filter

File: rotkehlchen/tests/unit/test_eth2.py [hunk 10]
```diff
@@ -761,3 +807,84 @@ def test_validator_daily_stats_with_db_interaction(  # pylint: disable=unused-ar
         )
         assert stats_call.call_count == 1
         assert stats[:len(expected_stats)] == expected_stats
+
+
+@pytest.mark.parametrize('default_mock_price_value', [FVal(1.55)])
+def test_validator_with_custom_ownership(  # pylint: disable=unused-argument  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_eth2.py:813]: no no no. Do not add another new test that creates a whole new api server, scrapes for multiple validators (this is super slow) just to test ownership. 

Add a small part in the test above to test it.

What you should test is marking the ownership as 0.45% (or whatever) of a validator after it has had its daily stats removed.

**Review by LefterisJP — COMMENTED** [review]
Quite a few thigns to change.

Also very important thing missing.

An endpoint with the ability to edit a validator. The only thing you can edit would be the ownership percentage (which is why it did not exist so far).

This should be tested ofc too

File: rotkehlchen/api/v1/resources.py (hunk 1 above)
Comment (LefterisJP) [reply rotkehlchen/api/v1/resources.py:1281]: I don't really see why you would change percentage to proportion. We use the term `weight` for the underlying assets. So I don't see the connection at all. Having different name in backend and frontend is more confusing.

File: rotkehlchen/api/v1/encoding.py (hunk 2 above)
Comment (yabirgb) [reply rotkehlchen/api/v1/encoding.py:2471]: Yes, I tested it here https://github.com/rotki/rotki/blob/e1eba5f80abda901160c8ffe9a38c336dcf9f32a/rotkehlchen/tests/api/test_eth2.py#L258

the tests makes the addition of the validators (with and without the `ownership_percentage`) and then compares to the serialization that contains the `ownership_proportion`

File: rotkehlchen/chain/manager.py [hunk 11]
```diff
@@ -1680,12 +1685,22 @@ def get_eth2_history_events(
             if stats_entry.pnl_balance.amount == ZERO:
                 continue
 
-            if stats_entry.pnl_balance.amount > ZERO:
+            # Take into account the validator ownership proportion if is not 100%
+            validator_ownership = validators_ownership.get(stats_entry.validator_index, FVal(1))
+            if validator_ownership != FVal(1):
+                pnl_balance = Balance(
```
Comment (LefterisJP) [inline rotkehlchen/chain/manager.py:1691]: nitpick: Not sure if we need it anywhere else, but this would be a good addition to the `Balance()` class. A multiplication by an int/Fval. Not really needed though  ... just a thought in case we do this in other places too.

File: rotkehlchen/chain/manager.py [hunk 12]
```diff
@@ -1710,10 +1725,26 @@ def get_eth2_validators(self) -> List[Eth2Validator]:
             raise ModuleInactive('Cant get eth2 validators since the eth2 module is not active')
         return DBEth2(self.database).get_validators()
 
+    def edit_eth2_validator(self, validator_index: int, ownership_proportion: FVal) -> None:
+        """Edit a validator to modify its ownership proportion. May raise:
+        - ModuleInactive if eht2 module is not active
```
Comment (LefterisJP) [inline rotkehlchen/chain/manager.py:1730]:
```suggestion
        - ModuleInactive if eth2 module is not active
```

File: rotkehlchen/db/eth2.py [hunk 13]
```diff
@@ -185,11 +197,19 @@ def add_validators(self, validators: List[Eth2Validator]) -> None:
...
         )
         self.db.update_last_write()
 
+    def edit_validator(self, validator_index: int, ownership_proportion: FVal) -> int:
+        cursor = self.db.conn.cursor()
+        cursor.execute(
+            'UPDATE eth2_validators SET ownership_proportion=? WHERE validator_index = ?',
+            (str(ownership_proportion), validator_index),
+        )
+        return cursor.rowcount
```
Comment (LefterisJP) [inline rotkehlchen/db/eth2.py:211]: This should always be either 1 or 0. Due to the primary key of the table it can't be more than 1. So here just simply raise `InputError()` with an error of editing non existing validator directly from this function.

Also add it and explain it the docstring.

File: rotkehlchen/chain/manager.py [hunk 14]
```diff
@@ -1710,10 +1725,26 @@ def get_eth2_validators(self) -> List[Eth2Validator]:
...
+        """
+        eth2 = self.get_module('eth2')
+        if eth2 is None:
+            raise ModuleInactive('Cant edit eth2 validators since the eth2 module is not active')
+        changes = DBEth2(self.database).edit_validator(
+            validator_index=validator_index,
+            ownership_proportion=ownership_proportion,
+        )
+        if changes != 1:
+            raise InputError(f'{changes} changes were made in the database.')
```
Comment (LefterisJP) [inline rotkehlchen/chain/manager.py:1741]: This error makes no sense, especially to a user who would see it. Tells them nothing about what went wrong or where. Just do what I said inside the `edit_validator()` function.

```suggestion
```

File: rotkehlchen/db/eth2.py [hunk 15]
```diff
@@ -169,7 +171,17 @@ def get_validator_daily_stats(
...
-        return [ValidatorDailyStats.deserialize_from_db(x) for x in results]
+        daily_stats = [ValidatorDailyStats.deserialize_from_db(x) for x in results]
+        # Take into account the proportion of the validator owned
+        validators_ownership = {
+            validator.index: validator.ownership_proportion
+            for validator in self.get_validators()
+        }
+        for i, daily_stat in enumerate(daily_stats):
+            owned_proportion = validators_ownership.get(daily_stat.validator_index, FVal(1))
+            if owned_proportion != ONE:
```
Comment (LefterisJP) [inline rotkehlchen/db/eth2.py:182]: Ah good. We have a constant for `ONE`. Why not use it also in all other places where you initialize `FVal(1)`?

File: rotkehlchen/db/eth2.py [hunk 16]
```diff
@@ -169,7 +171,17 @@ def get_validator_daily_stats(
...
+                daily_stats[i] = daily_stat._replace(pnl=daily_stat.pnl * owned_proportion)
```
Comment (LefterisJP) [inline rotkehlchen/db/eth2.py:183]: Argh this sucks. Well it's no longer an immutable named tuple with this functionality. So please change `ValidatorDailyStats` to a dataclass with frozen=False. Then you will no longer need to use the `_replace` hack

File: rotkehlchen/db/schema.py [hunk 17]
```diff
@@ -556,7 +556,8 @@
 DB_CREATE_ETH2_VALIDATORS = """
 CREATE TABLE IF NOT EXISTS eth2_validators (
     validator_index INTEGER NOT NULL PRIMARY KEY,
-    public_key TEXT NOT NULL UNIQUE
+    public_key TEXT NOT NULL UNIQUE,
+    ownership_proportion TEXT NOT NULL
```
Comment (LefterisJP) [inline rotkehlchen/db/schema.py:560]: You need to add this change in the table creation in v30_v31 DB upgrade

File: rotkehlchen/tests/api/test_eth2.py [hunk 18]
```diff
@@ -335,6 +339,53 @@ def test_add_get_delete_eth2_validators(rotkehlchen_api_server, start_with_valid
...
+            public_key='0x827e0f30c3d34e3ee58957dd7956b0f194d64cc404fca4a7313dc1b25ac1f28dcaddf59d05fbda798fa5b894c91b84fb',  # noqa: E501
+            ownership_proportion=FVal(0.4),
+        ),
+        Eth2Validator(
+            index=43948,
+            public_key='0x922127b0722e0fca3ceeffe78a6d2f91f5b78edff42b65cce438f5430e67f389ff9f8f6a14a26ee6467051ddb1cc21eb',  # noqa: E501
+            ownership_proportion=FVal(0.5),
+        ),
+    ]
+
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_eth2.py:355]: remove extra newline here
```suggestion
    ]
```

File: rotkehlchen/tests/api/test_eth2.py [hunk 19]
```diff
@@ -335,6 +339,53 @@ def test_add_get_delete_eth2_validators(rotkehlchen_api_server, start_with_valid
...
+    assert_simple_ok_response(response)
+
+    # Edit the second validor
+    response = requests.patch(
+        api_url_for(
+            rotkehlchen_api_server,
+            'eth2validatorsresource',
+        ), json={
+            'validator_index': 5235,
+            'ownership_percentage': 40,
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_eth2.py:375]: can you make the edit test to use a floating point? Shoudl also work. Something like '40.25' or whatever

File: rotkehlchen/tests/api/test_eth2.py [hunk 20]
```diff
@@ -335,6 +339,53 @@ def test_add_get_delete_eth2_validators(rotkehlchen_api_server, start_with_valid
...
+            'eth2validatorsresource',
+        ), json={
+            'validator_index': custom_percentage_validtors[1].index,
+            'public_key': custom_percentage_validtors[1].public_key,
+            'ownership_percentage': 50,
+        },
+    )
+    assert_simple_ok_response(response)
+
+    # Edit the second validor
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_eth2.py:368]:
```suggestion
    # Edit the second validator
```

**Review by LefterisJP — COMMENTED** [review]
Some smaller comments.

File: rotkehlchen/chain/ethereum/modules/eth2.py [hunk 21]
```diff
@@ -375,9 +382,14 @@ def add_validator(self, validator_index: Optional[int], public_key: Optional[Eth
...
 
                 raise RemoteError(f'Failed to query beaconcha.in for validator data due to: {msg}') from e  # noqa: E501
-
         # by now we have a valid index and pubkey. Add to DB
-        dbeth2.add_validators([Eth2Validator(index=valid_index, public_key=valid_pubkey)])
+        dbeth2.add_validators([
+            Eth2Validator(
+                index=valid_index,
+                public_key=valid_pubkey,
+                ownership_proportion=ownership_proportion,
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/eth2.py:390]: Hmmm did you remove the proportion changes in the get_balances() function? I can't find them. The get_balances() of eth2 should show us the balances in ETH2 per validator along with the proper percentage, right?

**Review by LefterisJP — COMMENTED** [review]
Some more stuff to fix

File: rotkehlchen/chain/ethereum/typing.py [hunk 22]
```diff
@@ -109,23 +110,24 @@ def __hash__(self) -> int:
...
-    missed_attestations: int = 0
-    orphaned_attestations: int = 0
-    proposed_blocks: int = 0
-    missed_blocks: int = 0
-    orphaned_blocks: int = 0
-    included_attester_slashings: int = 0
-    proposer_attester_slashings: int = 0
-    deposits_number: int = 0
-    amount_deposited: FVal = ZERO
+    missed_attestations: int = field(default=0)
```
[the remark this answers is not in the document]
Comment (yabirgb) [reply rotkehlchen/chain/ethereum/typing.py:122]: I didn't understand correctly the documentation. I believed that the right way to initialize default values was like this but is only needed if you need extra functionality from data classes. Will undo this

**Review by LefterisJP — APPROVED** [review]
Approving and will merge since the 2 things I see are nitpicks.

Also the test does not seem related to the changes ... I assume some CI problem

```
2021-12-26T12:55:38.8574106Z self = <rotkehlchen.db.dbhandler.DBHandler object at 0x7f04d03c9490>
2021-12-26T12:55:38.8574786Z 
2021-12-26T12:55:38.8575535Z     def update_last_write(self) -> None:
2021-12-26T12:55:38.8576151Z         # Also keep it in memory for faster querying
2021-12-26T12:55:38.8576738Z         self.last_write_ts = ts_now()
2021-12-26T12:55:38.8577290Z         cursor = self.conn.cursor()
2021-12-26T12:55:38.8577846Z         cursor.execute(
2021-12-26T12:55:38.8578679Z             'INSERT OR REPLACE INTO settings(name, value) VALUES(?, ?)',
2021-12-26T12:55:38.8579574Z             ('last_write_ts', str(self.last_write_ts)),
2021-12-26T12:55:38.8579994Z         )
2021-12-26T12:55:38.8580374Z >       self.conn.commit()
2021-12-26T12:55:38.8581144Z E       pysqlcipher3.dbapi2.OperationalError: disk I/O error
2021-12-26T12:55:38.8581729Z 
```
